[Feature] Scrollable tabs

fredriklindell opened the pull request:
This commit implements request in below issue:

https://github.com/callemall/material-ui/issues/1148

When device width is less than 992 tabs will paginate if the tab
container width is wider than device sreen width.

alitaheri commented:
Wow very nice @fredriklindell :+1: :+1: Thanks a lot. any chance you could also add an example in the docs site explicitly for this behavior?

oliviertassinari commented:
@fredriklindell That looks like good start.
Yes, could you add a new example of this so people know how to use it and to prevent regression.
Could you also fix the regression on the existing examples?
![capture d ecran 2016-01-10 a 14 51 12](https://cloud.githubusercontent.com/assets/3165635/12221810/f5f54720-b7a9-11e5-8495-ad1ef7a5189a.png)
Thanks!

fredriklindell commented:
@oliviertassinari ofc i can fix a new example. Not sure what to add though, since the implementation works for above on small screens. I could add an example with more tabs - i'll do that :) what do you mean with regression, do you mean that the tab should stretch to fill the width? I could change to default prop for stretch to true, now it is false, should I do that?

mbrookes commented:
Sorry if I'm misunderstanding, but is this a regression in the examples, or a regression in the tabs component? I.e if it changes the layout of the existing example, is it going to break current deployments? If so it's the tabs code that needs to be fixed, not the example. (Obviously an example for the paged tabs is needed.

fredriklindell commented:
@oliviertassinari alright, thank you for the explaination. I changed the Tabs default prop to stretch by default. However doing so I noticed that the examples break. Due to that the widths of the html elements in componentDidMount is not correct. If a resize is triggered everything is working, it is just the initial state. No sure if it has to do with the CodeExample component. I have to investigate a bit more. Do you have any ideas? :)

I also added an example with 12 tabs.

oliviertassinari commented on src/tabs/tabs.jsx:
```diff
@@ -81,25 +112,35 @@ const Tabs = React.createClass({
   getDefaultProps() {
     return {
       initialSelectedIndex: 0,
+      stretchTabContainer: false,
```
should be `true` by default to not introduce breaking changes.

oliviertassinari commented:
> I also added an example with 12 tabs.

That sounds like a good idea :+1:.

fredriklindell commented:
@oliviertassinari Correct, I have changed it to true.

But before I commit I need to fix the initial rendering error i am seeing in the examples. The widths read in componentWillMount is not correct. When I try this in my app it works. I have tried wrapping the code in my app with the ClearFix component - but it still works.

Any ideas? If you try by yourself you will notice that the ink-bar is not rendered correct. I think I need some help with this.

oliviertassinari commented:
> Any ideas?

I see that you have pushed th code, I will try to have a look.

fredriklindell commented:
@oliviertassinari and regarding the possibility to change back to true in 0.15 sounds good, because the calculated width when stretching the tabs might not be equal to the width of the actual DOM element, thus making the ink-bar a bit offline to the tab.

fredriklindell commented:
@oliviertassinari I have uploaded a new commit which somewhat fixes the problem. However, the strange thing is that the DOM elements are rendered with different widths between page reloads, it switched between two values with a rate of 50%. As I said I do not see these problems in my application only in the example page.

fredriklindell commented:
@oliviertassinari have you had any time to look at the PR?

Thanks in advance.

vishalvijay commented:
@fredriklindell Great work!!
@oliviertassinari When will this feature get released, waiting for it.

alitaheri commented:
@fredriklindell We are very sorry it took so long. But there were a lot of things we had to take care of. There were a bunch of other features added to tabs so this needs rebasing. If you don't have the time to work on this, i can take it from here. Thanks for working on this :+1:

fredriklindell commented:
@oliviertassinari Any update on this PR? Do I need to compliment anything?

Thanks in advance!

fredriklindell commented:
@alitaheri No worries, I will see what I can do tonight :)

oliviertassinari commented:
@fredriklindell Sorry for not catching up earlier. I have noticed various issues with this PR.
- The existing examples shouldn't change from what is now and independently of the screen size.
- Having a look at the specification here (https://www.google.com/design/spec/components/tabs.html#tabs-usage). I think that it would be good to add a `scrollable` boolean property. It would be `false` by default to not break the existing examples. When `true`, the width of the tab won't try to expand.
- I feel like it's highly increasing the complexity of the `Tabs` and `Tab` components. I'm wondering if we can't move it to a new component. (That remember me this PR https://github.com/callemall/material-ui/pull/1321).
- The scrollable feature doesn't work on mobile devices, this don't have to be implemented in this PR, but I'm wondering how we would implement it once this PR is merged.
- I think that it would be great to have this feature, I guess this can be done in another PR:
![janv 30 2016 12 57](https://cloud.githubusercontent.com/assets/3165635/12695521/3c9d1eda-c751-11e5-875d-6e4e2de78b4a.gif)

oliviertassinari commented on src/tabs/tabs.jsx:
```diff
@@ -167,6 +224,134 @@ const Tabs = React.createClass({
+  },
+
+  handleWindowWidthChange() {
+    this.replaceState(this.getNewState());
```
Can we not use `replaceState`?
> This method is not available on ES6 class components that extend React.Component. It may be removed entirely in a future version of React.

oliviertassinari replied:
Can we not use `replaceState`?
> This method is not available on ES6 class components that extend React.Component. It may be removed entirely in a future version of React.

oliviertassinari commented on src/tabs/tab-paginator-button.jsx:
```diff
@@ -0,0 +1,141 @@
+    let themeVariables = this.state.muiTheme.tabs;
+    const styles = {
+      buttonStyle: {
+        display: display ? '' : 'none',
```
When `display` if `false`, can we skip the rendering of the virtual dom?

oliviertassinari commented on src/tabs/tab.jsx:
```diff
@@ -97,12 +103,13 @@ const Tab = React.createClass({
     } = this.props;
 
-    const styles = this.mergeStyles({
+    const styles = this.prepareStyles({
```
`mergeStyles` was better. `prepareStyles` should only be called before providing it to the `style` property.

fredriklindell commented:
@oliviertassinari thanks for your answer, and no worries!

I have changed the code according to your comments above.

1. Not sure if I agree, due to when you have many tabs with long texts.  On small screen devices, the tab containers are divided equally between the screen size but the label text are not, which will end up with texts outside the device screen.
2. I think the scrollable effect should occur if the width of all tabs are wider that the screen width - always! Do you want the tabs to always stretch without taking the device screen in consideration?
3. I tend to disagree. It it still Tabs/Tab component, the complexity should not result in a new component if you ask me, since the complexity is needed for the scrollable function.
4. What?! What's not working? I have tried it via `npm run browser:development` and everything seems to work. Also my [demo](http://localhost:9005/#/tabs-demo) page.
5. I agree, get the current PR merged then a new PR for the functionality.

mbrookes commented:
On point 3. My concern is if we or another contributor decides to implement the overflow dropdown feature, (or some other capability) Tabs will be even more complex, and require checks to disable one or the other (or multiple) features.

Implementing as a HOC wrapper as per the SelectableEnhance for List allows the user to chose whether they need this capability, keeps the code modular and reduces complexity.

Please don't take this as a knock - you've done a fantastic job, we just need to keep in mind how this fits with the bigger picture.

newoga commented:
@fredriklindell Thanks a lot for your work on this feature. It's really cool.

>  On point 3, my concern is if we or another contributor decides to implement the overflow dropdown feature, (or some other capability) Tabs will be even more complex, and require checks to disable one or the other (or multiple) features.

I agree with this. And to clarify what @oliviertassinari said, I don't think we're suggesting create a different `<Tabs />` component, just a high order component that controls the scrolling behavior.

oliviertassinari commented:
@fredriklindell Yeah, sorry for behind rude. Given the popularity the project has quickly acquired, I focus more on increasing the code quality / fixing bugs than on implementing new features (later).
I really see two features in this PR:
 - The option to make the `Tab` *not expand* with different width. I feel like it should be the default for wide screen.
 - The *overflow* implementation for no-touch devices that show change depending on the size of the screen.

1. On small screen like mobile devices, I think that the `expand` and the  `no-expand` are two valid use case.
 - `expand`:
![capture d ecran 2016-02-01 a 12 52 32](https://cloud.githubusercontent.com/assets/3165635/12716703/bca510dc-c8e2-11e5-95d3-373bad10b3fd.png)
 - `no-expand` with overflow:
![capture d ecran 2016-02-01 a 12 54 11](https://cloud.githubusercontent.com/assets/3165635/12716728/f6e54190-c8e2-11e5-80b8-535c96fcbdc4.png)

1. > I think the scrollable effect should occur if the width of all tabs are wider that the screen width - always!

I agree.

1. One of the reasons I'm saying that, is because now, we would render two `TabPaginatorButton` component even if we don't need it. That was a negative impact on the performances.

1. It's working, but on small devices, the specification tells use to not use arrow but to let users scroll with their fingers.

fredriklindell commented:
@oliviertassinari, sorry for my late reply. 

You were not rough, sorry if I made that impression. It is just that I do not know about you future plans, which makes it a little harder setting the architecture of the feature. If you ask me, this PR implementation is better that what is in the repo at the moment. I would gladly help you further develop it, I just do not want the PR to grow to be to big to handle.

3. Is fixed! Not rendering if display is false.

4. I followed that Angular JS implementation of material ui. Is it crucial to this PR or something that could be changed in a new PR?

mbrookes commented:
@fredriklindell:
> It is just that I do not know about you future plans

We've recently tidied up the [roadmap](https://github.com/callemall/material-ui/blob/master/ROADMAP.md) which should help.

Ndpnt commented:
:+1:

vishalvijay commented:
@oliviertassinari Will this PR get merged?

oliviertassinari commented:
@vishalvijay We have agreed on splitting this PR into two chunks:
 - The first one is to implement a new property`fillWidth` with a `false` as the default value to keep the existing behavior.
 - The second one for the implement the overflow behavior.

vishalvijay commented:
@oliviertassinari Somebody working on the same or I can give a try?

oliviertassinari commented:
@vishalvijay Said he is off for two weeks. If you can give it try, that would be awesome :+1:.

alitaheri commented:
I'm continuing this work. Nobody else pick this up please :sweat_smile:

fredriklindell commented:
Hi @alitaheri, when I left for vacation three weeks ago there where problem with the new tab implementation using enhanced button. I think that, if not already solved, we should have a look on that component and why the text is allowed to be wider that the button. If you ask me, the button size should adjust to the text width, just like a normal html button does. I highlighted this problem, but have not had the time to look intuit any further. Not sure what has happened in the main report since then, probably a lot :-)

alitaheri commented:
@fredriklindell Yeah, a lot has changed. that's why I'm going to break down your PR into smaller ones and see if I can extract the logic into another component. This is a great feature :+1: I really need it in my app :sweat_smile: So I'll take it from here :+1:

nathanmarks commented:
@mbrookes This is **PR: On Hold** but is on the `0.15.0` release milestone, which isn't ideal :smile:  What's the actual status?

alitaheri commented:
@nathanmarks It's on hold because I'm working on it and this is just a place holder. I'm sure I can get it in the 0.15.0 release. :grin:

nathanmarks commented:
@alitaheri Ok, awesome! :+1:

aviraldg commented:
@alitaheri What's the progress on this? We need it in our app, so I wouldn't mind helping if you need it :smile:

alitaheri commented:
@aviraldg We do too, but I got a bit caught up with life, I plan on continuing the work on this in a couple of days. Although if you need this fast there are a couple of areas that should be fixed before this feature can be cleanly implemented. 

First of all, The InkBar should not rely on the assumption that all tabs are of the same size, it should rely on the calculated width of each tab. After this, Tabs will need to be patched to work with the InkBar.

fillWidth for the Tabs should be implemented.

And then another component that wraps the Tabs should be implemented to provide the logic for this feature. it's not easy to migrate this after all the changes that happened. That's why it will probably need a re-implementation.

I would spend time to get this in. But recently I got too busy with life.

nathanmarks commented:
@alitaheri Gonna finish this or want to close?

alitaheri commented:
Man this was so hard to migrate after all the change that I couldn't even understand the old code. I gave up on it after I got busy. We still need this feature in our app though -_- I guess I'll have to implement it from scratch when we get there.
